### ISSUE TITLE
Emerging Entity Detection

### ISSUE THREAD

**Issue by ZhaofengWu** [zalandoresearch/flair#199]
The readme mentions that this framework achieves 50.20 F1 on WNUT-17 dataset for Emerging Entity Detection, but is there any way I can reproduce this result? I looked over the [How to Reproduce Experiments](https://github.com/zalandoresearch/flair/blob/master/resources/docs/EXPERIMENTS.md) document but didn't see this task mentioned.

Or is there a way I can train some model for this task using this framework?

Thanks!

**Comment by alanakbik** [comment]
Hi @ZhaofengWu - I think we just used the standard parameters to train this model, so we never performed any parameter selection. I would probably start with a configuration in which you use the 'crawl' embeddings for English, paired with mixed language models forward and backward. Like this:

```python
embedding_list = [
    WordEmbeddings('crawl'),
    CharLMEmbeddings('mix-forward'),
    CharLMEmbeddings('mix-backward'),
]

embeddings: StackedEmbeddings = StackedEmbeddings(embeddings=embedding_list)

# initialize sequence tagger
from flair.models import SequenceTagger

tagger: SequenceTagger = SequenceTagger(hidden_size=256,
                                        embeddings=embeddings,
                                        tag_dictionary=tag_dictionary,
                                        tag_type=tag_type)

# initialize trainer
trainer: ModelTrainer = ModelTrainer(tagger, corpus)

trainer.train('resources/taggers/wnut',
              learning_rate=0.1,
              mini_batch_size=32,
              train_with_dev=True,
              patience=3)
```

Then, you could try experimenting with learning rates and mini batch sizes to improve the results. Remember to do parameter selection against dev data, to train_with_dev then should be set to False.

**Comment by ZhaofengWu** [comment]
Thanks! What should I pass in for the `corpus` argument? For crawl, I see that you have
```python
# 1. get the corpus
corpus: TaggedCorpus = NLPTaskDataFetcher.fetch_data(NLPTask.ONTONER)
print(corpus)

# 2. what tag do we want to predict?
tag_type = 'ner'

# 3. make the tag dictionary from the corpus
tag_dictionary = corpus.make_tag_dictionary(tag_type=tag_type)
```
But I'm not sure what I should do for W-NUT.

Also, what's the difference between `SequenceTaggerTrainer` and `ModelTrainer`? Why are we using the latter here?

**Comment by alanakbik** [comment]
Ah my mistake, not the `ModelTrainer`, you should use `SequenceTaggerTrainer`. 

(The ModelTrainer will be in the next release to unify the different trainer classes. But its not ready yet.)

For W-NUT, you need to define the column format and use the data fetcher to get the corpus. It will be something like this: 

```python
# define the column format of W-NUT - you might need to change this
columns = {1: 'text', 2: 'ner'}

corpus = NLPTaskDataFetcher.fetch_column_corpus(data_folder,
                                                          columns,
                                                          train_file='/path/to/train',
                                                          test_file='/path/to/test',
                                                          dev_file='/path/to/dev',
                                                          tag_to_biloes='ner') 
```
You might need to adapt the column definition so it fits the W-NUT data.

**Comment by ZhaofengWu** [comment]
Thanks!

**Comment by omerarshad** [comment]
does it worked for you? @ZhaofengWu

**Comment by Alexmac22347** [comment]
@ZhaofengWu Hey how did you get it to work? Ive only been able to get an f1 score of ~45% on the test set.
I tried running on, version 0.3.2, and version 0.2.1 from pip, but can't get to 50.20% f1.
I also quickly tried checking out branch release-0.4, but couldn't get the code to run.

**Comment by alanakbik** [comment]
@Alexmac22347 @ZhaofengWu  @omerarshad 

Try running with the following stacked embeddings: 

```python
embeddings = StackedEmbeddings(embeddings=[
    WordEmbeddings('crawl'),
    WordEmbeddings('twitter'),
    CharLMEmbeddings('news-forward'),
    CharLMEmbeddings('news-backward'),
])
```

With this, I get 49.5% on the current release-0.4 branch with default settings, averaged over 5 runs.

(We had the twitter embeddings in there in earlier versions, but removed them because we thought noone was using them. I just opened issue #291 to put them back in release-0.4.) Still, they should be there in release 0.2 if you want to use them now. 

@Alexmac22347 What was the problem with the release-0.4 branch? We're gearing up for a pip release so would be curious to hear what errors you were seeing.

**Comment by alanakbik** [comment]
@Alexmac22347  @ZhaofengWu here is the full code to reproduce experiments. 

1. Check out the current release-0.4 branch (or wait a week until we've pushed it to `pip`)

1. Run this code: 

```python
from flair.data_fetcher import NLPTaskDataFetcher, NLPTask
from flair.data import TaggedCorpus
from flair.embeddings import  WordEmbeddings, StackedEmbeddings, CharLMEmbeddings

# 1. get the corpus
from flair.trainers.trainer import ModelTrainer

corpus: TaggedCorpus = NLPTaskDataFetcher.load_corpus(NLPTask.WNUT_17)
print(corpus)
# 2. what tag do we want to predict?
tag_type = 'ner'

# 3. make the tag dictionary from the corpus
tag_dictionary = corpus.make_tag_dictionary(tag_type=tag_type)
print(tag_dictionary.idx2item)

embeddings = StackedEmbeddings(embeddings=[
    WordEmbeddings('crawl'),
    WordEmbeddings('twitter'),
    CharLMEmbeddings('news-forward'),
    CharLMEmbeddings('news-backward'),
])

# initialize sequence tagger
from flair.models import SequenceTagger

tagger: SequenceTagger = SequenceTagger(hidden_size=256,
                                        embeddings=embeddings,
                                        tag_dictionary=tag_dictionary,
                                        tag_type=tag_type)
print(tagger)
# initialize trainer
trainer: ModelTrainer = ModelTrainer(tagger, corpus)

results = trainer.train('resources/taggers/wnut-experiment',
                        max_epochs=150,
                        train_with_dev=True)

print(results)
```

We ran this code 5 times and got an average F1 of **49.49 +- 0.75** (standard deviation). 

As you can see, the standard deviation is quite high. This may suggest that the annealing is too hard. I.e. you could try increasing the annealing rate from 0.5 to 0.95 and lowering the patience from 3 to 0. Learning will then typically take longer, but the results become more stable (any typically even a bit better).

**Comment by Alexmac22347** [comment]
@alanakbik Hey, I was able to reproduce your results on release-0.4, the biggest thing was using the twitter embeddings. Btw the problem I had was because I forgot to install pytorch-pretrained-bert, it's all working now :)
